### DISCUSSION TITLE
[RFC] Animations v2

### DISCUSSION THREAD

**Discussion by natew** [tamagui/tamagui#553] [Ideas]
Right now animations work well for simple cases but don't let you do things you can do in CSS like loops, iterations, and timing-based animations (types only). Beyond fixing the types for timing based animations, we want to handle iterative animations.

Looking around I can see Framer Motion uses `transition` property, something like:

```tsx
<motion.circle
  cx={500}
  animate={{ cx: [null, 100, 200] }}
  transition={{ duration: 3, times: [0, 0.2, 1] }}
/>
```

Moti adopted it similary:

```tsx
<MotiView
  from={{ opacity: 0 }}
  animate={{ opacity: 1 }}
  transition={{
    repeat: 4,
    // when false, animation goes 0 -> 1, then starts over back at 0
    repeatReverse: false,
  }}
/>
```

But Framer and Moti also supports arrays for any given value to sequence:

```tsx
<MotiView
  animate={{
    scale: [0.1, 1.1, 1],
  }}
/>
```

Or customizing each step:

```tsx
<MotiView
  animate={{
    scale: [
      // you can mix primitive values with objects, too
      { value: 0.1, delay: 100 },
      1.1,
      { value: 1, type: 'timing', delay: 200 },
    ],
  }}
/>
```

For Tamagui I'd like to: support loops/sequences, stay close to CSS, and keep things simple. Supporting arrays on every value feels a bit heavy, but I could be persuaded. We already support a sort of array syntax too on animation props that is for scoping animation to specific properties:

```tsx
<Square
  animation={[
    'quick',
    {
      opacity: {
        overshootClamping: true,
      },
    },
  ]}
/>
```

The first attribute acts as the default - all properties animate with `quick`, and then the second lets you override per-property, so opacity will be `quick` but also add `overshootClamping: true`.

It also has repeat support:

```tsx
export default () => (
  <YStack
    animation={[
      // all attributes get "bouncy"
      'bouncy',
      // these are customized
      {
        y: 'slow',
        scale: {
          type: 'fast',
          repeat: 2,
        },
      },
    ]}
  />
)
```

So in fact seems we should already support repetition like this, and we can add loops here too:

```tsx
export default () => (
  <YStack
    animation={{
        loop: true,
        type: 'timing',
        delay: 100,
      }}
  />
)
```

The only final question then is sequence animations. I'm open to suggestions. The array syntax on property values to me is practically not an option, the types are too hard and it's too confusing, especially as many seem to like array values as useful for media query shorthands which we don't support but would nonetheless be easy to confuse.

It's likely something we just leave out for "v2" and just iterate on getting timing/loop in place as that is straightforward. Note that sequence would never be paired with loop, likely, at least Moti doesn't support and likely for some good reason.

It's possible we could add a new property:

```tsx
export default () => (
  <YStack
    animateSequence={[{ y: 0 }, { y: 100 }, { y: 10 }]}
  />
)
```

But maybe confusing when paired with `animation` that is an array but not mapping to the same sequence.

Or we can:

```tsx
export default () => (
  <YStack
    animation={{
      type: 'sequence',
      value: [{ y: 0 }, { y: 100 }]
    }}
  />
)
```

But how does that reconcile with the default style, and enterStyle.

**Comment by nandorojo** [comment]
> Note that sequence would never be paired with loop, likely, at least Moti doesn't support and likely for some good reason.

I think I ended up adding support for loops + sequences in a later version, I may not have updated the docs for it. Hard to remember at this point. But I'm pretty sure it works now. If only Reanimated weren't broken on @expo snack, I'd show an example.

I think that's a pretty unusual need though.

**Comment by kzlar** [comment]
Small though, if its not too difficult, an "onAnimationDone" prop could be useful in certain cases. Maybe its a v2.1 thing though.
Regardless, excited for this!

**Comment by redbar0n** [comment]
According to Fernando, the animation features you get in Moti over Tamagui (transitions) now are:

1. variants
2. dynamic transitions
3. sequences
4. on-the-fly animation configs

It'd be nice to integrate that into Tamagui, to minimize bundle size and dependencies, and to have a unified API.

FWIW, my top wishes for animation isn't loops, iterative animations, or sequences (which all seem to be aimed at hard core animators / fine-grained control), but... a simple API to:

- Increase speed of transitions dynamically. Slow transitions are nice for new users but quickly become annoying. So it'd be nice to be able to speed up transitions in line with how much the user has been using the app. Maybe this goes under point 2. or 4.
- Keeping certain elements on screen while navigating, to keep the contextual flow for the user. Aka.  shared element transitions: https://github.com/tamagui/tamagui/discussions/104 I understand if this is not first priority, since may be advanced to implement smoothly across platforms... but I think many apps could benefit from it, and a simple API for it would be another really great boon for people to use Tamagui.
- A simple declarative API for shake animations could be nice and find frequent use too. But it could beneficially be on an even higher/abstracted level than [shake animations in Moti](https://twitter.com/FernandoTheRojo/status/1470179831610548228), imho (could encourage more frequent use).

That's just my 2 cents and wishes (as an animation noob), FWIW. You might find that giving more low-level control may enable people to create those abstractions more easily themselves, and if so it might be the better way to go, since it could be the best of both worlds.

**Reply by nandorojo** [reply]
Moti also has full support for reanimated shared values, derived values, with native gesture handler support and worklets. I don’t see Tamagui ever fully matching up the feature set. Nor do I think it’s necessary.
